### PR TITLE
Trn 290 race detect

### PR DESCRIPTION
**Pull request by pchristopher1275**
https://jira2.iparadigms.com/browse/TRN-290

Ran race detector on code-base and made alterations to remove all data races. Mostly this entailed replacing object-level uses of WaitGroup with Semaphore. But also had to decouple StartDispatcher from the initialization of Dispatcher.

### CODE REVIEW

**Comment by pchristopher1275** [comment]
The code doesn't test Reset, that's a good point. Very easy to add, i'll do so. Not sure I need to test Done, as it is implemented as Add(-1). Not sure I agree with the assessment that " ... big test that wants to test everything ....". The test creates several threads, and has each thread use the class, concurrently. It tests Add and Wait (the two most important methods) in a concurrent environment.

**Comment by fitzboy** [comment]
question (on the test): how do you know that you are actually testing the concurrency well? I mean, you have no way to tell if the broadcast is being sent or if the mutexes are actually blocking and things are waiting for each other... I don't want to be picky, but given how I have seen my fair share of race conditions, I don't see anything here testing race conditions at all, I just see a basic functional test... like, I can't tell if semaphore will actually work well if one is waiting on the other... nothing in the test ensures that...

**Comment by pchristopher1275** [comment]
First off, in terms of how stressed the code actually is (whether mutexes are blocking etc.), that is admittedly info I don't have ready access to. In terms of real data-races, I can confirm that there are no races in the code in-so-much as the go race detector didn't report any. In order to give the race detector the best chance of finding races, I need to stress the code using concurrent access. 

BTW: i'm using Concurrent, as is used in golang. Specifically, in the go universe a distinction is made between being concurrent, and being parallelized. A piece of code can be neither of these, one of these, or both of these depending on code/configuration. But I know that the go scheduler is interleaving the execution of the goRoutineCount routines -- that's how the their user-space scheduler works. And that's all I need to reveal a race.

File: semaphore/semaphore.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+func New() *Semaphore {
+	s := &Semaphore{}
+	s.cond = sync.NewCond(&s.lock)
+	return s
+}
+
+func (sm *Semaphore) Reset() {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	sm.count = 0
```
Comment (fitzboy) [inline semaphore/semaphore.go]: can we add a condition in here such that if count > 0 then we definitely should log an ERROR right? i.e., someone is reseting a semaphore that other people are currently waiting on, that seems wrong.

File: semaphore/semaphore.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	sm.cond.Broadcast()
+}
+
+func (sm *Semaphore) Add(i int) {
```
Comment (fitzboy) [inline semaphore/semaphore.go]: is there a reason why you have it take in an int? I don't know when you would ever want to have it be not zero... seems like a lot of places in the code above you have Add(1) and as a reader of code, I wonder "what does 1 mean"? Would read easier if it was just 1...

File: semaphore/semaphore_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"time"
+)
+
+func TestReset(t *testing.T) {
+	semaphore := New()
+	semaphore.Add(100)
+	done := make(chan int)
+	go func() {
+		semaphore.Reset()
+		semaphore.Wait() // After a reset the count should be zero, so Wait() will return immediately.
```
Comment (fitzboy) [inline semaphore/semaphore_test.go]: I think that this test is not very clear. In reading it, why did you Add(100)? why not just Add(1)? further, in reading the test, when you call Reset() and then Wait() afterwards, it is not very clear either (though the comment explains it)... then why all the extra channel and then why base the test on a timeout? What if the machine is loaded or something, seems very non-deterministic... I would have write something more along the lines of:
semaphore := New()
semaphore.Add()
semaphore.Reset()
if semaphore.wouldBlock() {
     t.Fatalf("semaphore that has been reset would block!")
}

Further, there should be other Unit tests like this:
func TestNew(t *testing.T) {
semaphore := New()
if semaphore.wouldBlock() {
     t.Fatalf("new semaphore would block!")
}
}

func TestAddIncrementsCount(t *testing.T) {
semaphore := New()
semaphore.Add()
if !semaphore.wouldBlock() {
     t.Fatalf("semaphore with count of 1 would not block!")
}
}

func TestDone(t *testing.T) {
semaphore := New()
semaphore.Add()
semaphore.Done()
if semaphore.wouldBlock() {
     t.Fatalf("semaphore called with Done would block!")
}
}

**Comment by dankinder** [comment]
Sorry I haven't really been keeping track of this pull, but just wondering why the go race detector is calling out the way we use WaitGroup? I saw the link to that golang-nuts thread but I didn't see how it applies here, seems like we're using WaitGroup about as canonically as the example on the docs do.

File: semaphore/semaphore.go (hunk 2 above)
Comment (pchristopher1275) [inline semaphore/semaphore.go]: I agree it's lame. But the principal reason that this class exists is to replace WaitGroup (http://golang.org/pkg/sync/#WaitGroup) in parts of the code where WaitGroup will flag a data-race. And WaitGroup uses this funky Add(int), Done() interface. I wanted it to be really easy to drop Semaphore in where someone had used WaitGroup.

**Comment by pchristopher1275** [comment]
In response to @dankinder: I don’t think I agree with you that we use WaitGroup in the canonical way. I would argue that the canonical example of wait group is
{code}
var wg WaitGroup
for i := 1; i < 10; i++ {
	wg.Add(1)
	go func(){
		//do stuff
		wg.Done()
	}
}
wg.Wait()
{code}

First thing we’re doing differently is we’re not using a stack var to hold wg, so we have something more like
{code}
//object scoped wg
type Foo struct {
	//…
	wg WaitGroup
}

// in a function
for i := 1; i < 10; i++ {
	obj.wg.Add(1)
	go func(){
		//do stuff
		obj.wg.Done()
	}
}
obj.wg.Wait()
{code}

The second thing that we’re doing differently is we’re putting the whole thing in a big loop, so we have something more like

{code}
//object scoped wg
type Foo struct {
	//…
	wg WaitGroup
}

// in a function
for {
	for i := 1; i < 10; i++ {
		// The second time around this loop, the race detector will complain
		obj.wg.Add(1) 
		go func(){
			//do stuff
			obj.wg.Done()
		}
	}
	obj.wg.Wait()
}
{code}

It appears that WaitGroup uses some atomic-var magic, which the race detector can’t sort out. In the posts I saw, the author of WaitGroup argued the class was correctly implemented, and should flag a data-race in the way we’re using it. But Semaphore uses mutex/condition variables which the race detector knows will protect the read/writes, and so doesn’t flag the problem. 

It appears based on what we know right now, that WaitGroup should only really be used when the wg can be used in the first form.

**Comment by dankinder** [comment]
Huh, that's interesting. I'm wondering if the race condition would be fixed if we just did `obj.wg = sync.WaitGroup{}` right before we use it? That is, if the enclosing loop is the problem, I would think that would satisfy the race detector.

**Comment by dankinder** [comment]
Sorry if this is repetitive with your conversation with @cgilling about it, I'm just curious.

**Comment by pchristopher1275** [comment]
In response to @fitzboy:

Ok, let me take your argument point-by-point

FITZ: I think that this test is not very clear. In reading it, why did you Add(100)? why not just Add(1)?

RESPONSE: I’m setting the readers expectations. If I said Add(1), then it might appear to the user that Reset() is the same as Done(). But in fact, Reset() will clear any count, not just a count of 1.


FITZ: further, in reading the test, when you call Reset() and then Wait() afterwards, it is not very clear either (though the comment explains it)... 

RESPONSE: Not sure I can agree with this sentiment. The comment reads “After a reset the count should be zero, so Wait() will return immediately.” Since the users sees only a single Add in the test, either Wait() should return immediately as the comment says, or something wrong happened. 


FITZ: then why all the extra channel and then 

RESPONSE: I use a single channel to flag success of the test. I have to communicate back to the calling thread that Wait() returned. The channel is allocated, written too a single time, and then read from a single time. That seems pretty easy to understand. The only other option I really have is to use a single boolean, but in that case I MUST protect that boolean with a mutex, or the race detector will flag it as a problem. Channels are easier to use, and read, than mutex: so I used a channel.


FITZ: why base the test on a timeout? 

RESPONSE: If the comment that I highlight above weren’t true, the go-routine will hang forever. I need a way to end the test.


FITZ: What if the machine is loaded or something, seems very non-deterministic... 

RESPONSE: until a couple weeks back, very few of our tests where deterministic. This test is admittedly non-deterministic, but only in the weakest sense. The test is bound by 100ms, so the only real way it can fail errantly is if it takes > 100ms to call Reset(), Wait(), and then write to a done channel. Very unlikely. Especially compared to other timed tests we have, this one is very very unlikely to fail errantly.


FITZ: I would have write something more along the lines of:
….. code elided ….

RESPONSE: (a) In terms of code coverage, your tests are no better than mine — we both cover the same calls. (b) We have established,  in this project, that if we can avoid it (which we haven’t always), we do not add test-access methods like wouldBlock() (test-access method meaning a method that is present specifically to allow testing). We could add that to the interface, but it wasn’t needed to replicate the WaitGroup interface, and would have to have been made public as WouldBlock(). (c) Your suggested tests do not test the code concurrently. Hence, the race-detector won’t catch any data races — which, if you recall, is precisely the reason we’re writing this patch in the first place.

**Comment by pchristopher1275** [comment]
In response to @dankinder:

Yea, me an @cgilling chatted about this around the water cooler.

I tried your suggestion of obj.wg = sync.WaitGroup{}, doesn't work. The reason it doesn't work can be seen when you consider that obj.wg = sync.WaitGroup{} is really just zeroing the fields of obj.wg, which is, itself, a write to existing memory, and could induce a data race. When I spoke with gilling we talked about doing (the abominable) allocating a fresh WaitGroup on the heap per cycle. I presume this would fix the problem, at the cost of a memory allocation per cycle (and being incredibly silly).

**Comment by pchristopher1275** [comment]
BTW: the two comments from Chris that started this review, were from a previous pull request. I made any applicable changes, but github is out-of-wack and reposted them to this pr.

**Comment by dankinder** [comment]
Yeah little sad that we need this Semaphore class but I see the reasoning.

@fitzboy regarding the tests, I do think there is some sense to the way they are, since `TestConcurrent` would properly cause `go race` to flag a problem if there are race conditions in this class. For that reason I would keep the test and think it makes sense. The timeout also seems to make enough sense in this case, since in the general case (the code working properly) the test will actually finish very quickly.

However @pchristopher1275 I do agree with @fitzboy about covering the basic usecases of `Add`/`Wait`/`Done`, especially since they are part of the public API. I could imagine, for example, `Wait` not working properly and then the only test that would fail is `TestConcurrent`, and it would fail pretty mysteriously. For an API like this I would also expect `TestWait`, `TestDone`, etc. as verification of the basic individual parts. Or, since they are all interdependent (i.e. you need to use Add to test Wait anyway) a `TestBasic` or something that tests the functionality in it's simplest form and tries to inform us which step of the process went wrong. (Note: I do agree with you that we want to avoid creating private hooks for testing like wouldBlock, though a public method like `Count` or something may not be a bad idea anyway).

Outside of having some basic functionality test like that the code looks fine to me :facepunch:

**Pull request merged**